tslearn/nonmyopic.py

```diff
@@ -0,0 +1,509 @@
+            Delta_k = (average_distance - d_k) / average_distance
+            proba_current_cluster = 1 / (
+                math.exp(self.lamb * minimum_distance_to_average)
+                + math.exp(self.lamb * (minimum_distance_to_average - Delta_k))
+                + 10e-6
+            )
+            sk.append(proba_current_cluster)
+            sum_sk = sum_sk + sk[-1]
+        final_sk = [(x / sum_sk) for x in sk]
+        return final_sk
+
+    def cost_function(self, X, tau):
+        """
+        From a incomplete series xt, this function compute the expected cost of a prediction made at time "last time of
+        xt + tau"
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated for which we want to have the cost
+        tau: int
+            gives the future time for which the cost is calculated
+
+        Returns
+        --------
+        float : the computed cost
+        """
+        cost = 0
+        proba_clusters = self.get_cluster_prob(X=X)
+        truncated_t = X.shape[-1]
+        for current_cluster in range(0, self.number_cluster):
+            cost_y = 0
+            for y in range(0, self.number_classes_):
+                cost_y_hat = 0
+                for y_hat in range(0, self.number_classes_):
+                    if y != y_hat:
+                        cost_y_hat = (
+                            cost_y_hat
+                            + self.pyhatyck_["pyhatycks{0}".format(current_cluster)][
+                                truncated_t + tau - self.minimum_time_stamp
+                            ][y][y_hat]
+                        )
+                cost_y = cost_y + self.pyck_[y, current_cluster] * cost_y_hat
+            cost = cost + proba_clusters[current_cluster] * cost_y
+        cost = cost + self.cost_time(t=truncated_t + tau)
+
+        return cost
+
+    @staticmethod
+    def minimize_integer(end_of_time, function_to_minimize, xt):
+        """
+        We want to minimize a function "funct" according a time series "xt" for a list of integers from 0 to "stop"
+        Parameters
+        ----------
+        end_of_time: int
+            the highest integer at which the function to be minimized is calculated
+        function_to_minimize: funct
+            the function of interest
+        xt: vector
+            the time series related to the function "funct"
+
+        Returns
+         ------
+         int : The integer in {0,...,stop} that minimizes "funct"
+         float : the so fat minimum cost
+        """
+        tau_star = 0
+        so_far_minimum_cost = function_to_minimize(xt, tau_star)
+        for tau in range(1, end_of_time):
+            current_cost = function_to_minimize(xt, tau)
+            if current_cost < so_far_minimum_cost:
+                so_far_minimum_cost = current_cost
+                tau_star = tau
+        return tau_star + xt.shape[-1], so_far_minimum_cost
+
+    def classification(self, xt, cost_function):
```

Comment:
Do you sometimes use a `cost_function` that is different from `self.cost_function`? If not, then it should not be an argument of the function.

Reply:
Also, this method should have a more explicit name, something like `_predict_at_fixed_length` (underscore at the beginning to mark a private function, that should not be called from the outside of the class).

Reply:
I let the self.cost_function in case one might want to experiment new cost_function in the future. As it is for now it can be removed though.